File: Έγγραφα SEM4105/Λύσεις ασκήσεων/10. Δέκατο εργαστήριο/01/forward_euler.py
import math

def f(x,y):
    return math.cos(x)-x*math.sin(x)


def forward_euler(x0, y0, x1, f):
    return y0 + (x1-x0) * f(x0, y0)


		     
def correct(x):
    return 2.0 + x*math.cos(x)


def main():
    a = 0.0
    yinit = 2.0
    b = 3.0
    
    h = 0.01
    
    x0 = a
    y0 = yinit

    print(x0, y0, correct(x0))

    while x0 < b:
        if x0+h > b:
            h = b-x0

        y0 = forward_euler(x0, y0, x0+h, f)
        x0 = x0+h
        print(x0, y0, correct(x0))
      


main()

File: Έγγραφα SEM4105/Λύσεις ασκήσεων/10. Δέκατο εργαστήριο/01/forward_euler.c
#include <stdio.h>
#include <math.h>


double f(double x, double y)
{
  return cos(x)-x*sin(x);
}


double forward_euler(double x0, double y0, double x1,
		     double (*f)(double x, double y))
{
  return y0 + (x1-x0) * f(x0, y0);    
}
		     

double correct(double x)
{
    return 2.0 + x*cos(x);
}


int main()
{
  double const a = 0.0;
  double const yinit = 2.0;

  double const b = 3.0;
   

  double h = 0.01;

  double x0 = a;
  double y0 = yinit;

  printf("%g %g %g\n", x0, y0, correct(x0));

  while (x0 < b) {
    if (x0 +h > b) {
      h = b-x0;
    }
    
    y0 = forward_euler(x0, y0, x0+h, f);
    x0 += h;
    printf("%g %g %g\n", x0, y0, correct(x0));
  }
    
  return 0;
}
